File: fetchr/run/clean.py
from __future__ import annotations

from typing import Any, Dict, Tuple

import numpy as np
import pandas as pd

from .io_utils import normalize_series

_VALID_FILL_METHODS = {"none", "ffill", "bfill", "both", "time", "linear"}


def _norm(value: Any) -> str:
    return str(value or "").strip()


def _norm_l(value: Any) -> str:
    return _norm(value).lower()


def _to_numeric_series(series: pd.Series, *, name: str) -> pd.Series:
    out = pd.to_numeric(series, errors="coerce").copy()
    out.index = pd.to_datetime(out.index)
    out = out[~out.index.duplicated(keep="last")]
    out.sort_index(inplace=True)
    out.name = name
    return out


def _validate_winsor_quantiles(value: Any) -> Tuple[float, float] | None:
    if value is None:
        return None
    if not isinstance(value, (list, tuple)) or len(value) != 2:
        raise ValueError("winsor_quantiles must be a 2-item list [lower_q, upper_q]")
    lower = float(value[0])
    upper = float(value[1])
    if lower < 0.0 or upper > 1.0 or lower >= upper:
        raise ValueError("winsor_quantiles must satisfy 0 <= lower_q < upper_q <= 1")
    return lower, upper


def clean_series(task: Dict[str, Any], series: pd.Series, *, output_name: str) -> tuple[pd.Series, Dict[str, Any]]:
    s = _to_numeric_series(series, name=output_name)
    meta: Dict[str, Any] = {
        "winsorized_count": 0,
        "zscore_clipped_count": 0,
        "hampel_replaced_count": 0,
        "missing_before_fill": int(s.isna().sum()),
        "missing_after_fill": int(s.isna().sum()),
    }

    winsor = _validate_winsor_quantiles(task.get("winsor_quantiles"))
    if winsor is not None:
        lo_q, hi_q = winsor
        valid = s.dropna()
        if not valid.empty:
            lo = float(valid.quantile(lo_q))
            hi = float(valid.quantile(hi_q))
            before = s.copy()
            s = s.clip(lower=lo, upper=hi)
            changed = (before != s) & before.notna()
            meta["winsorized_count"] = int(changed.sum())
            meta["winsor_lower"] = lo
            meta["winsor_upper"] = hi

    zscore_threshold_raw = task.get("zscore_threshold")
    if zscore_threshold_raw is not None:
        threshold = float(zscore_threshold_raw)
        if threshold <= 0:
            raise ValueError("zscore_threshold must be > 0")
        valid = s.dropna()
        if not valid.empty:
            mu = float(valid.mean())
            sigma = float(valid.std(ddof=0))
            if sigma > 0:
                lo = mu - threshold * sigma
                hi = mu + threshold * sigma
                before = s.copy()
                s = s.clip(lower=lo, upper=hi)
                changed = (before != s) & before.notna()
                meta["zscore_clipped_count"] = int(changed.sum())

    hampel_window_raw = task.get("hampel_window")
    if hampel_window_raw is not None:
        window = int(hampel_window_raw)
        if window < 1:
            raise ValueError("hampel_window must be >= 1")
        n_sigma = float(task.get("hampel_n_sigma", 3.0))
        if n_sigma <= 0:
            raise ValueError("hampel_n_sigma must be > 0")
        min_periods = max(1, window // 2)
        rolling_med = s.rolling(window=window, center=True, min_periods=min_periods).median()
        mad = (s - rolling_med).abs().rolling(window=window, center=True, min_periods=min_periods).median()
        scale = 1.4826 * mad
        outlier_mask = (s - rolling_med).abs() > (n_sigma * scale)
        outlier_mask &= scale.notna() & (scale > 0)
        replace_mask = outlier_mask & rolling_med.notna()
        meta["hampel_replaced_count"] = int(replace_mask.sum())
        if bool(replace_mask.any()):
            s = s.where(~replace_mask, rolling_med)

    lower_bound = task.get("lower_bound")
    upper_bound = task.get("upper_bound")
    if lower_bound is not None or upper_bound is not None:
        lo = float(lower_bound) if lower_bound is not None else None
        hi = float(upper_bound) if upper_bound is not None else None
        if lo is not None and hi is not None and lo > hi:
            raise ValueError("lower_bound must be <= upper_bound")
        s = s.clip(lower=lo, upper=hi)

    smooth_window_raw = task.get("smoothing_window")
    if smooth_window_raw is not None:
        window = int(smooth_window_raw)
        if window < 1:
            raise ValueError("smoothing_window must be >= 1")
        s = s.rolling(window=window, min_periods=1).mean()

    fill_method = _norm_l(task.get("fill_method") or "none")
    if fill_method not in _VALID_FILL_METHODS:
        raise ValueError(f"fill_method must be one of {sorted(_VALID_FILL_METHODS)}")
    if fill_method == "ffill":
        s = s.ffill()
    elif fill_method == "bfill":
        s = s.bfill()
    elif fill_method == "both":
        s = s.ffill().bfill()
    elif fill_method == "time":
        s = s.interpolate(method="time", limit_direction="both")
    elif fill_method == "linear":
        s = s.interpolate(method="linear", limit_direction="both")

    meta["missing_after_fill"] = int(s.isna().sum())
    cleaned = normalize_series(s, name=output_name)
    meta["n_obs_in"] = int(series.shape[0])
    meta["n_obs_out"] = int(cleaned.shape[0])
    meta["fill_method"] = fill_method
    return cleaned, meta

File: econark-r/fetchr-R/run/clean.R
clean_series <- function(task, series_df, output_name) {
  s <- normalize_series_df(series_df, name = output_name)
  vals <- s$value
  meta <- list(
    winsorized_count = 0L,
    zscore_clipped_count = 0L,
    hampel_replaced_count = 0L,
    missing_before_fill = sum(is.na(vals)),
    missing_after_fill = sum(is.na(vals))
  )

  if (!is.null(task$winsor_quantiles)) {
    q <- as.numeric(unlist(task$winsor_quantiles))
    if (length(q) != 2 || q[1] < 0 || q[2] > 1 || q[1] >= q[2]) stop("winsor_quantiles must be [lower, upper]")
    lo <- stats::quantile(vals, probs = q[1], na.rm = TRUE, type = 7)
    hi <- stats::quantile(vals, probs = q[2], na.rm = TRUE, type = 7)
    before <- vals
    vals <- pmax(pmin(vals, hi), lo)
    meta$winsorized_count <- as.integer(sum(abs(before - vals) > 0, na.rm = TRUE))
    meta$winsor_lower <- as.numeric(lo)
    meta$winsor_upper <- as.numeric(hi)
  }

  if (!is.null(task$zscore_threshold)) {
    th <- as.numeric(task$zscore_threshold)
    mu <- mean(vals, na.rm = TRUE)
    sdv <- stats::sd(vals, na.rm = TRUE)
    if (is.finite(sdv) && sdv > 0) {
      lo <- mu - th * sdv
      hi <- mu + th * sdv
      before <- vals
      vals <- pmax(pmin(vals, hi), lo)
      meta$zscore_clipped_count <- as.integer(sum(abs(before - vals) > 0, na.rm = TRUE))
    }
  }

  if (!is.null(task$hampel_window)) {
    k <- as.integer(task$hampel_window)
    n_sigma <- ifelse(is.null(task$hampel_n_sigma), 3, as.numeric(task$hampel_n_sigma))
    n <- length(vals)
    repl <- 0L
    for (i in seq_len(n)) {
      lo <- max(1, i - k)
      hi <- min(n, i + k)
      w <- vals[lo:hi]
      med <- stats::median(w, na.rm = TRUE)
      mad <- stats::median(abs(w - med), na.rm = TRUE)
      scale <- 1.4826 * mad
      if (is.finite(scale) && scale > 0 && abs(vals[i] - med) > (n_sigma * scale)) {
        vals[i] <- med
        repl <- repl + 1L
      }
    }
    meta$hampel_replaced_count <- repl
  }

  if (!is.null(task$lower_bound)) vals <- pmax(vals, as.numeric(task$lower_bound))
  if (!is.null(task$upper_bound)) vals <- pmin(vals, as.numeric(task$upper_bound))

  if (!is.null(task$smoothing_window)) {
    w <- as.integer(task$smoothing_window)
    vals <- as.numeric(stats::filter(vals, rep(1 / w, w), sides = 1))
  }

  fill_method <- tolower(trimws(ifelse(is.null(task$fill_method), "none", as.character(task$fill_method))))
  if (fill_method == "ffill" || fill_method == "both") {
    for (i in seq_along(vals)) {
      if (is.na(vals[i]) && i > 1) vals[i] <- vals[i - 1]
    }
  }
  if (fill_method == "bfill" || fill_method == "both") {
    for (i in rev(seq_along(vals))) {
      if (is.na(vals[i]) && i < length(vals)) vals[i] <- vals[i + 1]
    }
  }
  if (fill_method == "linear" || fill_method == "time") {
    idx <- which(!is.na(vals))
    if (length(idx) >= 2) {
      vals <- stats::approx(x = idx, y = vals[idx], xout = seq_along(vals), method = "linear", rule = 2)$y
    }
  }

  out <- data.frame(date = s$date, value = vals, stringsAsFactors = FALSE)
  out <- normalize_series_df(out, name = output_name)
  meta$missing_after_fill <- sum(is.na(vals))
  meta$n_obs_in <- nrow(series_df)
  meta$n_obs_out <- nrow(out)
  meta$fill_method <- fill_method
  list(series = out, meta = meta)
}
